### PR TITLE
lineage-tracking

### PR DESCRIPTION
**Pull request by Swikar**
Lineage Tracking demoscript

kubeflow/kubeflow#5106

### CODE REVIEW

**Comment by Swikar** [comment]
@hamelsmu this is new PR https://github.com/kubeflow/blog/pull/34

**Comment by Swikar** [comment]
@hamelsmu To add further in this blog. Do I create new PR for that? /preview means?

**Comment by hamelsmu** [comment]
@jlewi this PR is ready for review, not sure what the process is or who is the SME.   There are some grammatical errors etc but I assume the designated reviewer can help with that

**Comment by jlewi** [comment]
@hamelsmu How come it doesn't show up in the preview?
https://deploy-preview-34--infallible-murdock-d05108.netlify.app/

@karlschriek @StefanoFioravanzo @aronchick I think you are all working on lineage so do you want to review this?

/hold
We should hold the blog post until after the 1.1 blog post because I think it depends on features in KF 1.1.

**Comment by jlewi** [comment]
@Swikar Thanks for doing this. It looks like the information on using metadata and lineage tracking is buried within the post. There is a lot of other stuff (using pipelines, fairing, training, prediction); that I think is distracting.

How about stripping all of that out and just focus on writing, reading from metadata?

Instead of linking to the YouTube Video can we provide people a link that would open up the lineage tracking UI in their cluster?

**Comment by jlewi** [comment]
@Swikar Could you please link to the relevant issue.

**Comment by hamelsmu** [comment]
Jeremy You have to use the /preview command and wait for the materialized
preview because the notebooks need to be converted first.  The netlify
hooks will not work in this case I would suggest turning the auto build off
if you can

**Comment by jlewi** [comment]
Thanks @hamelsmu

**Comment by jlewi** [comment]
@neuromage Can you suggest someone from metadata who can review this post?

**Comment by Swikar** [comment]
Changes made

**Comment by hamelsmu** [comment]
Why is there a README file in this PR?

image

**Comment by Swikar** [comment]
This will remove I believe upon merge.
As this doesn’t exist in my local repository. - Swikar/blog


Thanks,
Swikar

**Comment by karlschriek** [comment]
@Swikar @hamelsmu, I suppose I would qualify as an SME, and also since my opinion was asked somewhat further up I would be happy to review:

To be honest to me it isn't really that clear what this blog post is trying to tell me. Is the aim to explain what Workspaces, Runs and Executions are and how they work together? Why the need for the ModelServe class? Why is workspace wrapped as a @property on that class? There is also a lot of if/then logic going on. Could'd all of this just be written in a more imperative manner step by step so that the reader can follow what is happening?

Also where do Fairing and Seldon come it? They doesn't seem to be relevant here, but maybe I am missing something.

The code also isn't 100% runnable as you only show how to create the workspace, but not how to run the actual code. For a user actually follow the steps, `ws.list()` will return nothing.

This link is broken: https://github.com/kubeflow/metadata/blob/master/sdk/python/demo.ipynb

**Comment by hamelsmu** [comment]
Thanks Carl.  Yeah the link will be broken until the PR is merged you can
view the notebook in this fork though

**Comment by jlewi** [comment]
@karlschriek Thanks; I agree with your feedback. 

The intent of kubeflow/kubeflow#5106 was to highlight how people can use metadata to do lineage tracking. 

* Create a python script (or notebook) that uses the metadata SDK to write a sequence of executions and artifacts to TFX metadata
* Open up the lineage view in the KFP UI to view the lineage

@Swikar WDYT? Can we strip out all the unrelated bits (e.g. fairing, Seldon, etc....) and just focus on using the metadata SDK to create data which is then visible in the lineage tracking UI?

**Comment by Swikar** [comment]
Karl, Jeremy yes right. I am going to crafting more robust and simple way for audience to understand easily. Next version will be of that 1-2days


Thanks
Swikar

**Comment by Swikar** [comment]
/preview Jeremy, Karl please have a look. Any suggestion?

**Comment by jlewi** [comment]
@Swikar This is looking pretty great; thanks for the revisions. 

Could you please work to get this reviewed by @StefanoFioravanzo and @karlschriek since they are working on similar areas?

**Comment by Swikar** [comment]
> @Swikar This is looking pretty great; thanks for the revisions. 
> 
> 
> 
> Could you please work to get this reviewed by @StefanoFioravanzo and @karlschriek since they are working on similar areas?

@jlewi sure I will work with them to get reviewed.

**Comment by jlewi** [comment]
@StefanoFioravanzo @karlschriek ping? Will you be able to take a look and review the revisions?

**Comment by StefanoFioravanzo** [comment]
@jlewi I was on PTO the past week. I will review as soon as possible.

File: _notebooks/2020-07-28-lineage.ipynb
Comment (StefanoFioravanzo) [inline _notebooks/2020-07-28-lineage.ipynb:22]: ..to create **a** notebook

File: _notebooks/2020-07-28-lineage.ipynb
Comment (StefanoFioravanzo) [inline _notebooks/2020-07-28-lineage.ipynb:20]: Maybe remove the bullet list from this paragraph?

The intent of this ...
We are going to:
1. ...
2. ...
3. ...

File: _notebooks/2020-07-28-lineage.ipynb
Comment (StefanoFioravanzo) [inline _notebooks/2020-07-28-lineage.ipynb:24]: I would write this as:

"This blog post will first guide you through the metadata SDK API, to create and log several actions to the metadata DB. Afterwards, you will be able to navigate to the Kubeflow UI and the resulting lineage graph, which gives you a graphical representation of the dependencies between the objects you logged using the SDK"

Or something similar

File: _notebooks/2020-07-28-lineage.ipynb
Comment (StefanoFioravanzo) [inline _notebooks/2020-07-28-lineage.ipynb:77]: Create a new Workspace and Run

File: _notebooks/2020-07-28-lineage.ipynb
Comment (StefanoFioravanzo) [inline _notebooks/2020-07-28-lineage.ipynb:78]: A [Workspace](...) groups a set of pipelines or notebooks runs, and their related artifacts and executions

File: _notebooks/2020-07-28-lineage.ipynb
Comment (StefanoFioravanzo) [inline _notebooks/2020-07-28-lineage.ipynb:79]: [Store](...)  is an object that provides a connection to the Metadata gRPC service

File: _notebooks/2020-07-28-lineage.ipynb
Comment (StefanoFioravanzo) [inline _notebooks/2020-07-28-lineage.ipynb:80]: The [Run](...) object captures a pipeline or notebook run in a workspace

Not sure what you meant by "and group execution"

File: _notebooks/2020-07-28-lineage.ipynb
Comment (StefanoFioravanzo) [inline _notebooks/2020-07-28-lineage.ipynb:115]: I would explain more clearly what is the relationship between a Run and an Execution here. Maybe you want to say that an Execution is a specific "Instance" of a run? And that you can "bind" specific input/output artifacts to this instance?

File: _notebooks/2020-07-28-lineage.ipynb
Comment (StefanoFioravanzo) [inline _notebooks/2020-07-28-lineage.ipynb:143]: `DataSet` and `Model` are subclasses of `Artifact`. So the SDK provides some ready-to-use APIs for common Artifacts. These classes have specific fields (like `query` or `hyperparameters`). It would be nice to have a paragraph that describes what are the available Artifacts and if/how one could create new ones.

**Review by StefanoFioravanzo — CHANGES_REQUESTED** [review]

**Comment by StefanoFioravanzo** [comment]
@Swikar added some comments, mainly on wording and adding some more details. Overall looks pretty nice, thanks!!

**Comment by jlewi** [comment]
@karlschriek @StefanoFioravanzo I thought you were going to review this blog post since it relates to the model management and lineage tracking? It looks like this has been pending for quite a while.  Are there any major issues? Can we set an ETA for getting the review complete and the blog published?

@Swikar If you need to escalate and if possible; you might want to attend the model management meetings since @karlschriek hosts those.

/assign @karlschriek

**Comment by Swikar** [comment]
> @karlschriek @StefanoFioravanzo I thought you were going to review this blog post since it relates to the model management and lineage tracking? It looks like this has been pending for quite a while. Are there any major issues? Can we set an ETA for getting the review complete and the blog published?
> 
> @Swikar If you need to escalate and if possible; you might want to attend the model management meetings since @karlschriek hosts those.
> 
> /assign @karlschriek

@jlewi I made changes requested by StefanoFioravanzo. Waiting for approval from StefanoFioravanzo.

**Comment by StefanoFioravanzo** [comment]
@Swikar Thanks for addressing my comments

/lgtm

**Comment by jlewi** [comment]
@Swikar can you post a preview link to the blog, so we can do final approvals?

**Comment by jlewi** [comment]
LGTM except for the preview not showing up.

**Comment by Swikar** [comment]
Do you see preview?

**Comment by hamelsmu** [comment]
> LGTM except for the preview not showing up.

Jeremy - What do you mean by the preview not showing up?  I see it in https://github.com/kubeflow/blog/pull/34#issuecomment-698942585

**Comment by hamelsmu** [comment]
@jlewi keep in mind for fastpages, the netlify integration is not that helpful because it will not auto-convert notebooks, you have to rely on `/preview` instead.  Happy to explain this more if it is confusing

**Comment by jlewi** [comment]
@hamelsmu I mean that when I click on the link in #34 I don't actually see the blog post. Anyways I guess we can figure that out later.

/hold cancel
/lgtm
/approve

**Comment by jlewi** [comment]
This is live:
https://blog.kubeflow.org/jupyter/2020/07/28/lineage.html

Great job @Swikar

**Comment by Swikar** [comment]
> This is live:
> 
> https://blog.kubeflow.org/jupyter/2020/07/28/lineage.html
> 
> 
> 
> Great job @Swikar 

Thanks @jlewi

**Pull request merged**
